File: x/go/pb/any_test.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package pb_test

import (
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	. "github.com/synnaxlabs/x/testutil"

	"github.com/synnaxlabs/x/pb"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/wrapperspb"
)

var _ = Describe("Any", func() {
	Describe("AnyToPBAny", func() {
		It("Should return nil for nil input", func() {
			result, err := pb.AnyToPBAny(nil)
			Expect(err).ToNot(HaveOccurred())
			Expect(result).To(BeNil())
		})

		It("Should pass through an existing *anypb.Any", func() {
			original := MustSucceed(anypb.New(wrapperspb.String("hello")))
			result := MustSucceed(pb.AnyToPBAny(original))
			Expect(result).To(Equal(original))
		})

		It("Should pack a proto.Message directly", func() {
			msg := wrapperspb.String("hello")
			result := MustSucceed(pb.AnyToPBAny(msg))
			Expect(result.TypeUrl).To(ContainSubstring("StringValue"))
		})

		It("Should convert a map[string]any to structpb.Struct", func() {
			m := map[string]any{"key": "value", "count": float64(42)}
			result := MustSucceed(pb.AnyToPBAny(m))
			Expect(result.TypeUrl).To(ContainSubstring("Struct"))
		})

		It("Should convert an arbitrary struct via JSON marshaling", func() {
			type custom struct {
				Name  string `json:"name"`
				Value int    `json:"value"`
			}
			result := MustSucceed(pb.AnyToPBAny(custom{Name: "test", Value: 5}))
			Expect(result.TypeUrl).To(ContainSubstring("Struct"))
		})
	})

	Describe("AnyFromPBAny", func() {
		It("Should return nil for nil input", func() {
			result, err := pb.AnyFromPBAny(nil)
			Expect(err).ToNot(HaveOccurred())
			Expect(result).To(BeNil())
		})

		It("Should unpack a structpb.Struct to a map", func() {
			s := MustSucceed(structpb.NewStruct(map[string]any{
				"key": "value",
				"num": float64(42),
			}))
			packed := MustSucceed(anypb.New(s))
			result := MustSucceed(pb.AnyFromPBAny(packed))
			m, ok := result.(map[string]any)
			Expect(ok).To(BeTrue())
			Expect(m["key"]).To(Equal("value"))
			Expect(m["num"]).To(BeNumerically("==", 42))
		})

		It("Should return nil for an unregistered type URL", func() {
			packed := &anypb.Any{
				TypeUrl: "type.googleapis.com/unknown.Type",
				Value:   []byte{1, 2, 3},
			}
			result, err := pb.AnyFromPBAny(packed)
			Expect(err).ToNot(HaveOccurred())
			Expect(result).To(BeNil())
		})
	})

	Describe("Round Trip", func() {
		It("Should round-trip a map through ToPBAny and FromPBAny", func() {
			original := map[string]any{
				"name":   "test",
				"count":  float64(10),
				"nested": map[string]any{"inner": "data"},
			}
			packed := MustSucceed(pb.AnyToPBAny(original))
			result := MustSucceed(pb.AnyFromPBAny(packed))
			m, ok := result.(map[string]any)
			Expect(ok).To(BeTrue())
			Expect(m["name"]).To(Equal("test"))
			Expect(m["count"]).To(BeNumerically("==", 10))
			nested, ok := m["nested"].(map[string]any)
			Expect(ok).To(BeTrue())
			Expect(nested["inner"]).To(Equal("data"))
		})
	})
})

File: x/cpp/json/any_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "gtest/gtest.h"

#include "x/cpp/json/any.h"
#include "x/cpp/json/testutil/testutil.h"
#include "x/cpp/test/test.h"

namespace x::json {
TEST(ToAny, ObjectRoundTrip) {
    const json j = {{"name", "sensor"}, {"port", 8080}};
    const auto any = ASSERT_NIL_P(to_any(j));
    ASSERT_FALSE(any.type_url().empty());
    const auto result = ASSERT_NIL_P(from_any(any));
    ASSERT_EQ(result["name"], "sensor");
    ASSERT_EQ(result["port"], 8080);
}

TEST(ToAny, NullRoundTrip) {
    const auto any = ASSERT_NIL_P(to_any(json(nullptr)));
    ASSERT_TRUE(ASSERT_NIL_P(from_any(any)).is_null());
}

TEST(ToAny, NumberRoundTrip) {
    const auto any = ASSERT_NIL_P(to_any(json(42)));
    ASSERT_EQ(ASSERT_NIL_P(from_any(any)), 42);
}

TEST(ToAny, StringRoundTrip) {
    const auto any = ASSERT_NIL_P(to_any(json("sensor")));
    ASSERT_EQ(ASSERT_NIL_P(from_any(any)), "sensor");
}

TEST(ToAny, BoolRoundTrip) {
    const auto any = ASSERT_NIL_P(to_any(json(true)));
    ASSERT_EQ(ASSERT_NIL_P(from_any(any)), true);
}

TEST(ToAny, ArrayRoundTrip) {
    const auto any = ASSERT_NIL_P(to_any(json::array({1, 2})));
    const auto result = ASSERT_NIL_P(from_any(any));
    ASSERT_TRUE(result.is_array());
    ASSERT_EQ(result[0], 1);
    ASSERT_EQ(result[1], 2);
}

TEST(ToAny, ValueConversionErrorPropagates) {
    ASSERT_OCCURRED_AS_P(to_any(deeply_nested_object()), errors::VALIDATION);
}

TEST(FromAny, EmptyAnyReturnsNull) {
    ASSERT_TRUE(ASSERT_NIL_P(from_any(google::protobuf::Any())).is_null());
}

TEST(FromAny, PackedStructUnpacks) {
    google::protobuf::Struct s;
    ASSERT_NIL(to_struct(json{{"name", "sensor"}}, &s));
    google::protobuf::Any any;
    ASSERT_TRUE(any.PackFrom(s));
    ASSERT_EQ(ASSERT_NIL_P(from_any(any))["name"], "sensor");
}

TEST(FromAny, UnknownTypeReturnsError) {
    google::protobuf::Any any;
    any.set_type_url("type.googleapis.com/unknown.Type");
    ASSERT_OCCURRED_AS_P(from_any(any), errors::VALIDATION);
}

TEST(ToAny, NestedObjectRoundTrip) {
    const json j = {{"outer", {{"inner", "value"}}}, {"list", {1, 2, 3}}};
    const auto any = ASSERT_NIL_P(to_any(j));
    const auto result = ASSERT_NIL_P(from_any(any));
    ASSERT_EQ(result["outer"]["inner"], "value");
    ASSERT_EQ(result["list"][0], 1);
    ASSERT_EQ(result["list"][1], 2);
    ASSERT_EQ(result["list"][2], 3);
}
}
